suzieq/db/parquet/parquetdb.py

```diff
@@ -651,45 +651,65 @@ def _get_filtered_fileset(self, dataset: ds, namespace: list) -> ds:
         Returns:
             ds: pyarrow dataset of only the files that match filter
         """
-        if not namespace:
+        def check_ns_conds(ns_to_test: str, filter_list: List,
+                           cond: operator.or_) -> bool:
+            # pylint: disable=comparison-with-callable
+            res = not cond == operator.or_
+            for filter_val in filter_list:
+                res = cond(res, bool(re.search(
+                    f'namespace={filter_val}', ns_to_test)))
+            return res
+
+        if not namespaces:
             return dataset
 
-        newns = reduce_filter_list(namespace)
-        ns_filelist = []
-        chklist = dataset.files
-        for ns_filter in newns or []:
-            if ns_filter.startswith('!'):
-                ns_filter = ns_filter[1:]
-                use_not = True
-            else:
-                use_not = False
+        match_filters = []
+        not_filters = []
 
-            if ns_filter.startswith('~'):
-                ns_filter = ns_filter[1:]
-            else:
-                ns_filter = re.escape(ns_filter)
-
-            if use_not:
-                ns_filelist = \
-                    [file for file in chklist
-                     if (ns_section := next(
-                         (s for s in file.split('/')
-                          if s.startswith('namespace=')),
-                         None))
-                     and not re.search(f'namespace={ns_filter}', ns_section)]
-                # NOTs turn list from implicit OR into implicit AND
-                chklist = ns_filelist
+        for ns_match in namespaces:
+            if ns_match.startswith(('!~', '~!')):
```

Comment:
`~!` is no more a valid filter. if the user types `~!`, the `!` must be considered the first character of the regex